src/ast_editor_state.ml

```diff
@@ -1,53 +1,49 @@
 open Import
-module Uri = Vscode.Uri
+
+type ast_mode =
+  [ `Original_ast
+  | `Preprocessed_ast
+  ]
 
 type t =
-  { (* The webview store that maps string value of Uri.t of the document to the
-       related webview. *)
-    mutable webview_map : (string, WebView.t, String.comparator_witness) Map.t
-        (* Indicates current mode inside AST explorer (true = Original, false =
-           Prerocessed) *)
-  ; mutable original_mode : bool
-        (* Contains Some value when hover mode is enabled, None otherwise *)
+  { mutable webview_map : (string, WebView.t, String.comparator_witness) Map.t
+        (** The webview store that maps string value of Uri.t of the document to
+            the related webview. *)
+  ; mutable current_ast_mode : ast_mode
+        (** Indicates current AST mode inside AST explorer: original or
+            preprocessed *)
   ; mutable hover_disposable : Disposable.t option
-        (* A set containing original documents (as opposed to Preprocessed
-           Document) that have been changed since last Preprocessed Document
-           opening (or reload) *)
+        (** Contains Some value when hover mode is enabled, None otherwise *)
   ; mutable dirty_original_doc_set : (string, String.comparator_witness) Set.t
-        (* Mapping beween the original and Preprocessed Document used to
-           simultaneously communicate with a common webview. *)
+        (** A set containing original documents (as opposed to Preprocessed
+            Document) that have been changed since last Preprocessed Document
+            opening (or reload) *)
   ; mutable origin_to_pp_doc_map :
       (string, string, String.comparator_witness) Map.t
+        (** Mapping beween the original and Preprocessed Document used to
+            simultaneously communicate with a common webview. *)
   }
 
 let make () =
   let webview_map = Map.empty (module String) in
-  let original_mode = true in
-  let hover_disposable = None in
   let dirty_original_doc_set = Set.empty (module String) in
   let origin_to_pp_doc_map = Map.empty (module String) in
   { webview_map
-  ; original_mode
-  ; hover_disposable
+  ; current_ast_mode = `Original_ast
+  ; hover_disposable = None
   ; dirty_original_doc_set
   ; origin_to_pp_doc_map
   }
 
 let find_original_doc_by_pp_uri t uri =
-  let r =
-    Map.filter
-      ~f:(fun data -> String.equal (Uri.toString uri ()) data)
-      t.origin_to_pp_doc_map
-    |> Map.keys
-  in
-  match r with
-  | k :: _ -> Some k
-  | _ -> None
+  let uri_s = Uri.toString uri () in
+  Map.filter ~f:(fun data -> String.equal uri_s data) t.origin_to_pp_doc_map
+  |> Map.keys |> List.hd
```

Comment:
Why aren't we using `min_binding` or even `find_first` here?

Reply:
my first question as well -- because `Map` assumes that you find by a key (`val find : ('k, 'v, 'cmp) t -> 'k -> 'v option`), while here we search by a value

we could perhaps do as below to not filter the whole map but I'm not 100% sure it's much better

```ocaml
  t.origin_to_pp_doc_map |> Map.to_sequence
  |> Sequence.find ~f:(fun (_, data) -> String.equal uri_s data)
```

Reply:
I see. Then I suppose we should just iter or fold and raise once we find something. What we'd really like to avoid is just generating all these intermediate data structures

Reply:
You might want to double check if converting a map to a sequence is at all helpful. I think it also allocates the entire map in a list like data structure before it traverses with sequence

Reply:
Updated the implementation to short-circuit with a local exception. 

re Sequence
I thought the idea of a sequence is to lazily access elements

Reply:
The idea is there, but the implementation of to_seq in map leaves much to be desired:

```
    type 'a enumeration = End | More of key * 'a * 'a t * 'a enumeration

    let rec cons_enum m e =
      match m with
        Empty -> e
      | Node {l; v; d; r} -> cons_enum l (More(v, d, r, e))

    let rec seq_of_enum_ c () = match c with
      | End -> Seq.Nil
      | More (k,v,t,rest) -> Seq.Cons ((k,v), seq_of_enum_ (cons_enum t rest))

    let to_seq m =
      seq_of_enum_ (cons_enum m End)
```

As you can see, not much is gained by converting it to a Seq.